src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsReader.java

```diff
@@ -112,123 +108,259 @@ public SortedTermsReader(@Nonnull FileHandle termsDataFileHandle,
+            while (currentPointId < startingPointId)
+            {
+                currentPointId++;
+                readTerm(currentPointId, currentTerm);
+                currentBlockIndex = currentPointId >>> blockShift;
+            }
+
+            // Move forward to the ending point ID, returning the point ID if we find our term
+            while (currentPointId < endingPointId)
+            {
+                if (compareTerms(currentTerm, skipTerm) >= 0)
+                    return currentPointId;
+                currentPointId++;
+                if (currentPointId == meta.termCount)
+                    return -1;
+                readTerm(currentPointId, currentTerm);
+                currentBlockIndex = currentPointId >>> blockShift;
+            }
+            return endingPointId < meta.termCount ? endingPointId : -1;
+        }
+
+        @VisibleForTesting
+        public void reset() throws IOException
+        {
+            currentPointId = 0;
+            currentBlockIndex = 0;
+            termsInput.seek(termsDataFp);
+            readTerm(currentPointId, currentTerm);
         }
 
+
         @Override
         public void close()
         {
             termsInput.close();
         }
 
-        /**
-         * Advances the cursor to the next term and reads it into the current term buffer.
-         * <p>
-         * If there are no more available terms, clears the term buffer and the cursor's position will point to the
-         * one behind the last item.
-         * <p>
-         * This method has constant time complexity.
-         *
-         * @return true if the cursor was advanced successfully, false if the end of file was reached
-         * @throws IOException if a read from the terms file fails
-         */
-        @VisibleForTesting
-        protected boolean advance() throws IOException
+        private boolean notInCurrentBlock(long pointId, BytesRef term)
         {
-            if (pointId >= meta.termCount || ++pointId >= meta.termCount)
-            {
-                currentTerm.length = 0;
+            if (inLastBlock(pointId) || !peekNextBlock(pointId))
                 return false;
-            }
 
-            int prefixLength;
-            int suffixLength;
-            if ((pointId & TERMS_DICT_BLOCK_MASK) == 0L)
+            resetPosition();
+
+            return compareTerms(term, nextBlockTerm) >= 0;
+        }
+
+        private boolean inLastBlock(long pointId)
+        {
+            return pointId >>> blockShift == blockOffsets.length() - 1;
+        }
+
+        // Tries to load the starting value of the next block into nextBlockTerm. This will return false
+        // if the pointId is in the last block.
+        private boolean peekNextBlock(long pointId)
+        {
+            long blockIndex = (pointId >>> blockShift) + 1;
+
+            if (blockIndex >= blockOffsets.length())
+                return false;
```

Comment:
Trying to hit this line in a test, without success.

`notInCurrentBlock()` checks `inLastBlock()`...and then also checks that (albeit in an open-ended way) here in `peekNextBlock()`. Assuming we can do earlier validation to make sure we aren't trying to go to a block that doesn't exist, could we just rewrite this chunk of the code like this?

```
private boolean notInCurrentBlock(long pointId, BytesRef term)
{
    if (inLastBlock(pointId))
        return false;

    peekNextBlock(pointId);
    resetPosition();

    return compareTerms(term, nextBlockTerm) >= 0;
}
...
private void peekNextBlock(long pointId)
{
    long blockIndex = (pointId >>> blockShift) + 1;
    termsInput.seek(blockOffsets.get(blockIndex) + termsDataFp);
    readTerm(blockIndex << blockShift, nextBlockTerm);
}
```